### ISSUE TITLE
[Autoscaler] Request Resources semantics fix

### ISSUE THREAD

**Issue by wuisawesome** [ray-project/ray#11417]
### What is the problem?

`request_resources({"GPU": 1})` Should only launch a single gpu instance. Currently, it will continuously launch instances.

*Ray version and other system information (Python version, TensorFlow version, OS):*

### Reproduction (REQUIRED)
Please provide a script that can be run to reproduce the issue. The script should have **no external library dependencies** (i.e., use fake or mock data / environments):

If we cannot run your script, we cannot fix your issue.

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/latest/installation.html).

**Comment by PidgeyBE** [comment]
Hey @wuisawesome 
I cannot reproduce this on ray 1.0.
I use this script:
```
import os
import ray
from ray.autoscaler.commands import (  # ray.autoscaler.sdk.request_resources after 1.0
    request_resources,
)
head_service = os.environ.get('RAY_HEAD_SERVICE_HOST')
head_port = os.environ.get('RAY_HEAD_SERVICE_PORT_REDIS_PRIMARY')
ray.init(address=f"{head_service}:{head_port}")

request_resources(bundles=[{"GPU": 1}])
```
with this config:
```
    available_node_types:
      def_head:
          node_config: {}
          resources: {"CPU": 1}
          max_workers: 1
      def_worker:
          node_config: {}
          resources: {"CPU": 1, "GPU": 1}
          max_workers: 3
```
When I run the script, I see a worker is scaled up, I then get:
```
2020-10-16 08:41:49,973 INFO resource_demand_scheduler.py:69 -- Cluster resources: [{'object_store_memory': 33.0, 'memory': 97.0, 'node:10.208.0.245': 1.0, 'NO_GPU': 1000000.0, 'HEAD': 999999.0, 'CPU': 0.9}, {'GPU': 1.0, 'memory': 113.0, 'READY': 1000000.0, 'accelerator_type:RTX': 1.0, 'node:10.0.4.170': 1.0, 'CPU': 1.0, 'GPU:GeForce RTX 2080 Ti': 1.0, 'object_store_memory': 33.0}, {'node:10.208.0.34': 0.99, 'CPU': 0.9, 'memory': 1.0, 'object_store_memory': 1.0, 'accelerator_type:RTX': 1.0}]
2020-10-16 08:41:49,973 INFO resource_demand_scheduler.py:70 -- Node counts: defaultdict(<class 'int'>, {'def_head': 1, 'def_worker': 1, 'def_regulator': 1})
2020-10-16 08:41:49,973 INFO resource_demand_scheduler.py:73 -- Resource demands: [{'GPU': 1}]
2020-10-16 08:41:49,973 INFO resource_demand_scheduler.py:74 -- Unfulfilled demands: []
2020-10-16 08:41:49,973 INFO resource_demand_scheduler.py:80 -- Node requests: []
2020-10-16 08:41:49,997 INFO autoscaler.py:542 -- Cluster status: 1/0 target nodes (0 pending)
```

After `idle_timeout_minutes`, the worker is scaled down, the resource request shifts to `Unfulfilled demands` and a new worker is started again.
For me this is the expected behaviour.
If I then rum `request_resources(bundles=[{}])`, the worker gets scaled down after  `idle_timeout_minutes` and the cycle stops...

**Comment by ericl** [comment]
I think the issue here is the semantics of request_resources()-- whether it means request additional resources [including] vs [excluding] current resource usage. Currently it implements the latter I believe.

**Comment by wuisawesome** [comment]
Well it has historically implemented the former. We changed the semantics when we updated the resource demand scheduler to take into account resource usage

**Comment by PidgeyBE** [comment]
Indeed, currently (in ray 1.0) `request_resources()` lets you request additional resources excluding current resource usage.
This has one big downside I think, as it opens up a small gap for race conditions.
1. Imagine you request_resources for actor A, which requires 1 GPU. The autoscaler will scale up 1 GPU worker.
2. Once the worker is up, you can start actor A, but at this point `request_resources` still contains `{"GPU": 1}`  so once you start the actor, again an additional worker will be started. (which is not what you want)
To avoid this, the user should manually reset the requested resources _right before_ starting the actor. This is possible of course, but imagine you are in a situation where you have tens of actors to spawn. In this case again you can request resources for all of them, wait a bit till the resources become available and once the resources are available, reset the requested resources list right before starting all the actors.
In this case, if starting the actors takes longer than the `worker_idle_timeout`, I think you will run into troubles and workers will scale down again before the actors were actually deployed...
This can be mitigated of course by adapting the `worker_idle_timeout` to your application...

Just a thought :)
In case `request_resources()` would include current resource usage, the "issue" would not be present I think...

**Comment by PidgeyBE** [comment]
@ericl Why is this not a bug but an enhancement?
I've found another issue with it:

```
    available_node_types:
      def_head:
          node_config: {}
          resources: {"CPU": 1}
          max_workers: 1
      def_worker:
          node_config: {}
          resources: {"CPU": 1, "GPU": 1}
          max_workers: 3
```
If you have a config like this, and you do `request_resources(bundles=[{"GPU": 1}, {"GPU": 1}])`, the necessary resources will never be available!
The autoscaler will start with scaling up 1 worker (instead of 2):
```
2020-10-19 13:14:05,488 INFO resource_demand_scheduler.py:73 -- Resource demands: [{'GPU': 1}, {'GPU': 1}]
2020-10-19 13:14:05,488 INFO resource_demand_scheduler.py:74 -- Unfulfilled demands: [{'GPU': 1}]
2020-10-19 13:14:15,564 INFO resource_demand_scheduler.py:80 -- Node requests: []
2020-10-19 13:14:15,580 INFO autoscaler.py:542 -- Cluster status: 1/0 target nodes (0 pending)
```
 After `worker_idle_timeout` it will scale down the 1 worker again and scale up a new worker. But it will never scale 2 workers...

**Comment by PidgeyBE** [comment]
And another issue is that it seems that after autoscaling 1 worker, the system does not autoscale anymore for additional resources:

```
import os
import ray
from ray.autoscaler.commands import (  # ray.autoscaler.sdk.request_resources after 1.0
    request_resources,
)
head_service = os.environ.get('RAY_HEAD_SERVICE_HOST')
head_port = os.environ.get('RAY_HEAD_SERVICE_PORT_REDIS_PRIMARY')
ray.init(address=f"{head_service}:{head_port}")


@ray.remote(num_gpus=1)
class ActorA:
    def __init__(self):
        pass

# 1. Request resources
request_resources(bundles=[{"GPU": 1}])
```
```
# 2. WAIT FOR WORKER TO BE AUTOSCALED
request_resources(bundles=[{}])
a = ActorA.remote()
```
```
# 3.  WAIT FOR ACTOR TO BE DEPLOYED
# Request resources again
request_resources(bundles=[{"GPU": 1}])
```
```
# Nothing happens
request_resources(bundles=[{"GPU": 1},{"GPU": 1}])
```

Output of monitor.err:
```

2020-10-19 15:47:32,821 INFO resource_demand_scheduler.py:73 -- Resource demands: [{'GPU': 1}, {'GPU': 1}]
2020-10-19 15:47:32,821 INFO resource_demand_scheduler.py:74 -- Unfulfilled demands: [{'GPU': 1}, {'GPU': 1}]
2020-10-19 15:47:32,821 INFO resource_demand_scheduler.py:80 -- Node requests: []
2020-10-19 15:47:32,861 INFO autoscaler.py:542 -- Cluster status: 1/1 target nodes (0 pending)
```
Without intentions to scale more workers...

**Comment by ericl** [comment]
@wuisawesome can you take a look at the reported upscaling issue?

**Comment by wuisawesome** [comment]
Yeah cc @AmeerHajAli in case you happen to know off the top of your head.
